nacos/src/main/java/com/linecorp/armeria/internal/nacos/QueryInstancesClient.java

```diff
@@ -0,0 +1,223 @@
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.HttpEntity;
+import com.linecorp.armeria.common.QueryParamsBuilder;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+/**
+ * A Nacos client that is responsible for
+ * <a href="https://nacos.io/en-us/docs/v2/guide/user/open-api.html">Nacos Open-Api - Query instances</a>.
+ */
+final class QueryInstancesClient {
+
+    static QueryInstancesClient of(NacosClient nacosClient, @Nullable LoginClient loginClient,
+                                   String nacosApiVersion) {
+        return new QueryInstancesClient(nacosClient, loginClient, nacosApiVersion);
+    }
+
+    private final WebClient webClient;
+
+    @Nullable
+    private final LoginClient loginClient;
+
+    private final String nacosApiVersion;
+
+    QueryInstancesClient(NacosClient nacosClient, @Nullable LoginClient loginClient,
+                         String nacosApiVersion) {
+        webClient = nacosClient.nacosWebClient();
+        this.loginClient = loginClient;
+        this.nacosApiVersion = nacosApiVersion;
+    }
+
+    CompletableFuture<List<Endpoint>> endpoints(String serviceName, @Nullable String namespaceId,
+                                                @Nullable String groupName, @Nullable String clusterName,
+                                                @Nullable Boolean healthyOnly, @Nullable String app) {
+        requireNonNull(serviceName, "serviceName");
+        return queryInstances(serviceName, namespaceId, groupName, clusterName, healthyOnly, app)
+                .thenApply(response -> response.data.hosts.stream()
+                        .map(QueryInstancesClient::toEndpoint)
+                        .filter(Objects::nonNull)
+                        .collect(toImmutableList()));
+    }
+
+    CompletableFuture<QueryInstancesResponse> queryInstances(String serviceName, @Nullable String namespaceId,
+                                                 @Nullable String groupName, @Nullable String clusterName,
+                                                 @Nullable Boolean healthyOnly, @Nullable String app) {
+        requireNonNull(serviceName, "serviceName");
+        final StringBuilder path = new StringBuilder("/").append(nacosApiVersion).append("/ns/instance/list?");
+        final QueryParamsBuilder paramsBuilder = NacosClientUtil
+                .queryParamsBuilder(namespaceId, groupName, serviceName, clusterName, healthyOnly, app, null,
+                                    null, null);
```

Comment:
It would be inefficient to create `QueryParams` for the same parameters for all requests.
How about creating and caching the `QueryParams` as a member field when `QueryInstancesClient` is created?